api/nakadi-event-bus-api.yaml

```diff
@@ -606,22 +488,394 @@ paths:
           description: Partition id
           type: string
           required: true
-        - name: X-Flow-Id
-          in: header
+        - $ref: '#/parameters/XFlowId'
+      responses:
+        '200':
+          description: OK
+          schema:
+            $ref: '#/definitions/Partition'
+        '401':
+          description: Client is not authenticated
+          schema:
+            $ref: '#/definitions/Problem'
+
+  /subscriptions:
+    put:
+      tags:
+        - managed-api
+      security:
+        - oauth2: ['nakadi.event_stream.read']
+      description: |
+        This endpoint creates a subscription for EventTypes. The subscription is needed to be able to
+        consume events from EventTypes in a high level way when Nakadi stores the offsets and manages the
+        rebalancing of consuming clients.
+        The subscription is identified by its key parameters (owning_application, event_types, consumer_group). If
+        this endpoint is invoked several times with the same key subscription properties in body (order of even_types is
+        not important) - the subscription will be created only once and for all other calls it will just return
+        the subscription that was already created.
+      parameters:
+        - name: subscription
+          in: body
+          description: Subscription to create
+          schema:
+            $ref: '#/definitions/Subscription'
+          required: true
+      responses:
```

Comment:
> But we are not sure if we should always reply with 201 and remove 200 from the API since there would be no means for the client to figure out if the subscription was just created of if it already existed

Now that it's a POST method, I went back through the http spec.

- 201 for created is fine.

- 200 for existing is fine to model this as a no-op request as per the documentation above

If you wanted to indicate that the response returned in either case does not require a subsequent GET request to fetch the subscription data, ie the response body is up to date, that can be done by sending a Content-Location header as well.

I also recommend making this an absolute URI and not a relative one. That allows domain level redirection which could be useful for a scaled out system. Also, the latest version of HTTP allows a relative URL (because it represents modern browser capabilities), but the old version required an absolute URL. So I would stick to absolute for now. I'll guess this is documented as relative as a implementation convenience (because it doesn't require the code to be configured for its domain) but I don't think that's worth it; it's easier for clients to work with absolute values.

Based on that, here's some possible text:

```
          headers:
             Location:
               description: | 
               	The absolute URI for this subscription resource. 
               type: string
               example: 'https://example.org/api/subscriptions/105a76d8-db49-4144-ace7-e683e8f4ba46'
             Content-Location:
               description: |
               	If the Content-Location header is present and the same as the Location header
               	the client can assume it has an up to date representation of the 
               	Subscription and a corresponding GET request is not needed.
               type: string
               example: 'https://example.org/api/subscriptions/105a76d8-db49-4144-ace7-e683e8f4ba46'
```